packages/dynamo-store/source/src/DynamoStoreSource.mts

```diff
@@ -274,65 +269,35 @@ function inflate(event: ITimelineEvent<EventBody>): ITimelineEvent {
 }
 
 export class DynamoStoreSource {
-  limiter: LimitFunction
-
+  private inner: TailingFeedSource
   constructor(
     private readonly index: AppendsIndex.Reader,
     private epochs: AppendsEpoch.Reader.Service,
     private readonly options: Omit<CreateOptions, "context">,
   ) {
-    this.limiter = pLimit(options.maxConcurrentStreams)
-  }
-
-  async handleTranche(
-    client: DynamoStoreSourceClient,
-    partition: AppendsPartitionId,
-    signal: AbortSignal,
-  ) {
-    let pos = Checkpoint.ofPosition(
-      await this.options.checkpoints.load(
-        this.options.groupName,
-        AppendsPartitionId.toString(partition),
-      ),
-    )
-    if (pos === Checkpoint.initial && this.options.startFromTail) {
-      pos = await Impl.readTailPositionForTranche(this.index, this.epochs, partition)
-    }
-    while (!signal.aborted) {
-      for await (const batch of client.crawl(partition, pos)) {
-        const streams = new Map<StreamName, ITimelineEvent[]>()
-        for (const [stream, event] of batch.items) {
-          const events = streams.get(stream)
-          if (events) {
-            events.push(inflate(event))
-          } else {
-            streams.set(stream, [inflate(event)])
-          }
-        }
-
-        const promises = []
-        for (const [stream, events] of streams) {
-          promises.push(this.limiter(() => this.options.handler(stream, events)))
-        }
-
-        await Promise.all(promises)
-        if (batch.checkpoint !== pos) {
-          await this.options.checkpoints.commit(
-            this.options.groupName,
-            AppendsPartitionId.toString(partition),
-            batch.checkpoint,
-          )
-          pos = batch.checkpoint
-        }
-        if (batch.isTail) await sleep(this.options.tailSleepIntervalMs, signal).catch(() => {})
-      }
+    if (!this.options.categories && !this.options.categoryFilter) {
+      throw new Error("Either categories or categoryFilter must be specified")
     }
+    const sink = new StreamsSink(this.options.handler, this.options.maxConcurrentStreams)
+    const categoryFilter = options.categoryFilter ?? ((c) => options.categories!.includes(c))
```

Comment:
in cosmos I recently added a stream level filter - if there is no other usage of the category name in the lower level, better to map categories/categoryFilter/steamFilter to a stream predicate lambda at this level (though default surface API of a list of categories as a prominent default continues to be the main line)

Reply:
could be more efficient to predicate that way 🤔 

```ts
streamName.startsWith(categoryName) && streamName[categoryName.length] === '-'
```

Reply:
interesting, I guess for the average category list, doing a few of these even beats splitting(and alocating) and then doing a set membership check

Reply:
<img width="965" alt="Screenshot 2023-09-16 at 14 55 07" src="https://github.com/equinox-project/equinox-js/assets/3721521/ca81f480-29f0-4f3b-9eee-3c78c9fb56e5">

So as expected, parsing the category is expensive, and a set with 2elems is slower than a linear search. Using a streamName filter with a for loop doing `startsWith(category + "-")` outperforms both by a factor of 2

Reply:
Moved to stream filter

Reply:
nice. Should do same in F# (wonder whether your initial algorithm with the startswith and checking the char after the prefix is `-` might be more efficient there)...